Add back custom Auth api

cdevienne opened the pull request:
Add back a SetCustomClientAuth as a replacement for SetClientAuthMethod that was removed for 1.0.x.)

Resolves #476 (waiting for #434)
 
/cc @nats-io/core

kozlovic commented on server/auth.go:
```diff
@@ -53,6 +70,14 @@ func (p *Permissions) clone() *Permissions {
 }
 
+func (s *Server) SetCustomClientAuth(auth Auth) {
+	s.customClientAuth = auth
```
I believe that we should protect this and make sure one can set it only before `server.Start()` has been invoked and return and error otherwise. 
Without locking/protection, any one could call this function (same for `SetCustomRouterAuth`) at any point when server is running and that would produce a race if code is trying to authorize a client.

One approach would be to protect setting this with `s.mu.Lock()/Unlock()` and check for `s.running`. If true, it means `Start()` has been invoked and therefore this call should fail (should return `error`).

A test case using those 2 new APIs (with dummy custom auth) would be welcome, including situations trying to set custom auth after server has been started.

cdevienne commented:
Thanks for the feedback. I will modify my PR.

About the validity of adding it back or not, I expected this.
But I really need it, so I did a patch and hoped it would influence somehow the discussions (hence the PR).

kozlovic commented:
@cdevienne We appreciate the contribution! I just wanted to make sure that it was understood that my comments were about the code itself and not a green light on the PR itself at this point.

ColinSullivan1 commented:
This sounds good to me since #434 is not implemented yet.  While the previous API wasn't exposed or really documented, it was public, and a few users were using it.  Thank you for the contribution!

cdevienne commented:
I just pushed a new version of my branch.
To avoid the need for security checks, I added the CustomClientAuth and CustomRouterAuth as server.Options attributes.
This way we benefit from the protection given by SetOpts.

The test I added make sure CustomClientAuth works as expected. I did not add a test for CustomRouterAuth though, as I did not have an easy way the setup a cluster in the test. I will happily take a hint for doing it in a simple way (simpler that routes_test.go).

kozlovic commented on server/opts.go:
```diff
@@ -71,6 +71,9 @@ type Options struct {
 	TLSConfig      *tls.Config   `json:"-"`
 	WriteDeadline  time.Duration `json:"-"`
+
+	CustomClientAuth Auth `json:"-"`
```
Could we rename those to `CustomClientAuthentication` (same for Router) to remove ambiguity between Authentication and Authorization (permissions)?

kozlovic commented on server/server_test.go:
```diff
@@ -386,3 +386,34 @@ func TestNilMonitoringPort(t *testing.T) {
 }
+
+type DummyAuth struct {
+	count int
```
Would remove this...

kozlovic commented on server/server_test.go:
```diff
@@ -386,3 +386,34 @@ func TestNilMonitoringPort(t *testing.T) {
+}
+
+func (d *DummyAuth) Check(c ClientAuth) bool {
+	d.count++
```
Would remove that and make it return false always, or if you want to make test more sophisticated, you could check `c.GetOpts().Username` and return value based on that.

kozlovic commented on server/server_test.go:
```diff
@@ -386,3 +386,34 @@ func TestNilMonitoringPort(t *testing.T) {
+
+	if clientAuth.count != 1 {
+		t.Error("Client auth should have been delegated to CustomClientAuth")
+	}
```
With changes proposed above, that would simplify and check that `Connect()` actually fails. Here is a modified version along with Router check:
```
func TestCustomClientAuth(t *testing.T) {
	opts := DefaultOptions()
	opts.CustomClientAuth = &DummyAuth{}

	s := RunServer(opts)
	defer s.Shutdown()

	addr := fmt.Sprintf("nats://%s:%d", opts.Host, opts.Port)
	if _, err := nats.Connect(addr); err == nil {
		t.Fatal("Expected client to fail to connect")
	}
}

func TestCustomRouterAuth(t *testing.T) {
	opts := DefaultOptions()
	opts.CustomRouterAuth = &DummyAuth{}
	opts.Cluster.Host = "127.0.0.1"
	s := RunServer(opts)
	defer s.Shutdown()
	clusterPort := s.ClusterAddr().Port

	opts2 := DefaultOptions()
	opts2.Cluster.Host = "127.0.0.1"
	opts2.Routes = RoutesFromStr(fmt.Sprintf("nats://127.0.0.1:%d", clusterPort))
	s2 := RunServer(opts2)
	defer s2.Shutdown()
	if nr := s2.NumRoutes(); nr != 0 {
		t.Fatalf("Expected no route, got %v", nr)
	}
}
```

cdevienne commented:
Thanks @kozlovic for the helpful review.
I think its ready now.

kozlovic commented on server/server_test.go:
```diff
@@ -386,3 +386,44 @@ func TestNilMonitoringPort(t *testing.T) {
+type DummyAuth struct{}
+
+func (d *DummyAuth) Check(c ClientAuthentication) bool {
+	return c.GetOpts().Username != "test"
```
When I was suggesting using `c.GetOpts().Username` was if you wanted to be more sophisticated and have a test that tries to connect with an allowed user name and one without.

cdevienne replied:
done

pward123 commented:
Is the custom auth's "RegisterUser" method ever called?

cdevienne commented:
@pward123 RegisterUser is not called in the tests I added, but I do use it in my project. It is important to keep it.
